pkgs/net-lib/net/server.rkt

```diff
@@ -0,0 +1,100 @@
+                      #:accept-proc [accept tcp-accept]
+                      #:close-proc [close tcp-close]
+                      #:timeout-evt-proc [make-timeout-evt (λ (_thd _in _out _break-sent?) never-evt)])
+  (define can-break?
+    (break-enabled))
+  (define paramz
+    (current-parameterization))
+  (define listener
+    (listen port max-allow-wait reuse? host))
+  (define server-thd
+    (thread
+     (lambda ()
+       (dynamic-wind
+         void
+         (lambda ()
+           (with-handlers ([exn:break? void])
+             (let loop ([in-progress 0])
+               (loop
+                (with-handlers ([exn:fail:network?
+                                 (λ (e)
+                                   (begin0 in-progress
+                                     ((error-display-handler)
+                                      (format "Connection error: ~a" (exn-message e))
+                                      e)))])
+                  (sync/enable-break
+                   (handle-evt
+                    (thread-receive-evt)
+                    (lambda (_)
+                      (let drain-loop ([in-progress in-progress])
+                        (if (thread-try-receive)
+                            (drain-loop (sub1 in-progress))
+                            in-progress))))
+                   (handle-evt
+                    (if (< in-progress max-concurrent) listener never-evt)
+                    (lambda (l)
+                      (define client-cust (make-custodian))
+                      (parameterize ([current-custodian client-cust])
+                        (parameterize-break #f
+                          (define-values (in out)
+                            (accept l))
+                          (define client-thd
+                            (thread
+                             (lambda ()
+                               (call-with-parameterization
+                                paramz
+                                (lambda ()
+                                  (break-enabled can-break?)
+                                  (parameterize ([current-custodian (make-custodian client-cust)])
+                                    (handle in out)))))))
+                          (thread
+                           (lambda ()
+                             (sync client-thd (make-timeout-evt client-thd in out #f))
+                             (when (thread-running? client-thd)
+                               (break-thread client-thd)
+                               (sync client-thd (make-timeout-evt client-thd in out #t)))
+                             (thread-send server-thd 'done void)
+                             (custodian-shutdown-all client-cust)))
+                          (add1 in-progress)))))))))))
+         (lambda ()
+           (close listener))))))
+  (lambda ()
+    (break-thread server-thd)
+    (thread-wait server-thd)))
+
+(define run-server
+  (make-keyword-procedure
+   (lambda (kws kw-args . args)
```

Comment:
It would be nice to get `object-name`, `procedure-arity`, and `procedure-keywords` right here.

Reply:
I can fix the procedure name and arity right here, but is there a way to fix `procedure-keywords` without spelling out all the keywords?

Reply:
You can use procedure-reduce-keyword-arity.

Reply:
Thanks!